src/firecracker/examples/uffd/uffd_utils.rs

```diff
@@ -377,109 +566,70 @@ impl Runtime {
-                        let mut current_pos = 0;
-                        let mut exit_loop = false;
-
-                        loop {
-                            // Read more data into the buffer if there's space
-                            let mut iov = [libc::iovec {
-                                iov_base: (buffer[current_pos..]).as_mut_ptr() as *mut libc::c_void,
-                                iov_len: buffer.len() - current_pos,
-                            }];
-
-                            if current_pos < BUFFER_SIZE {
-                                let ret = unsafe { self.stream.recv_with_fds(&mut iov, &mut fds) };
-                                match ret {
-                                    Ok((0, _)) => break,
-                                    Ok((n, 1)) => current_pos += n,
-                                    Ok((n, 0)) | Ok((_, n)) => panic!("Wrong number of fds: {}", n),
-                                    Err(e) if e.errno() == libc::EAGAIN => {
-                                        if exit_loop {
-                                            break;
-                                        }
-                                        continue;
+                        while let Some(msg) = uffd_msg_iter.next() {
+                            match msg {
+                                UffdMsgFromFirecracker::Mappings(mappings) => {
+                                    let (guest_memfd, userfault_bitmap_memfd) =
+                                        if uffd_msg_iter.secret_free {
+                                            (
+                                                Some(unsafe {
+                                                    File::from_raw_fd(uffd_msg_iter.fds[1])
+                                                }),
+                                                Some(unsafe {
+                                                    File::from_raw_fd(uffd_msg_iter.fds[2])
+                                                }),
+                                            )
+                                        } else {
+                                            (None, None)
+                                        };
+
+                                    let uffd_handler = UffdHandler::from_mappings(
+                                        mappings,
+                                        unsafe { File::from_raw_fd(uffd_msg_iter.fds[0]) },
+                                        guest_memfd,
+                                        userfault_bitmap_memfd,
+                                        self.backing_memory,
+                                        self.backing_memory_size,
+                                    );
+
+                                    let fd = uffd_handler.uffd.as_raw_fd();
+                                    if uffd_handler.guest_memfd.is_some() {
+                                        self.main_handler_fd = Some(fd);
                                     }
```

Comment:
I would move this entire mappings retrival logic out of the event loop and back into the uffd handler construction path. Since we now read the FDs from the socket without reading any json, the iterator doesn't really need to know about fds at all. They can just be read before we even construct the iterator. As for the mappings, we can set `<UffdMsgIterator as Iterator>::Item` to be `FaultRequest`, and simply directly call `recv_mappings()` from the uffd handler construction code